File: 315__Count_of_Smaller_Numbers_After_Self.py
# Binary Search Tree Solution
class Node:
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.count = 1
        self.left_tree_size = 0


class Solution(object):
    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        root = Node(float('inf'))
        result = []

        for i in range(len(nums)-1, -1, -1):
            root, count = self.insert(nums[i], root)
            result.append(count)

        return result[::-1]


    def insert(self, val, node):
        if not node:
            node = Node(val)
            return node, 0

        if val == node.val:
            node.count += 1
            return node, node.left_tree_size

        if val < node.val:
            node.left_tree_size += 1
            node.left, count = self.insert(val, node.left)
            return node, count

        node.right, count = self.insert(val, node.right)

        return node, node.count + node.left_tree_size + count



# Binary Index Tree Solution


class Solution(object):
    def countSmaller(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        if not nums or not len(nums): return []

        rank = {num: i+1 for i, num in enumerate(sorted(nums)) }
        N = len(nums)
        res = []
        BIT = [0] * (N+1)

        def update(i):
            while i <= N:
                BIT[i] += 1
                i += i & -i

        def query(i):
            sum_ = 0
            while i > 0:
                sum_ += BIT[i]
                i -= i & -i
            return sum_

        for num in nums[::-1]:
            res.append(query(rank[num] - 1))
            update(rank[num])

        return res[::-1]

File: Java/315__Count_of_Smaller_Numbers_After_Self.java
class Solution {

    class TreeNode {
        TreeNode left;
        TreeNode right;
        int val;
        int count = 1;
        public TreeNode(int val) {
            this.val = val;
        }
    }

    public List<Integer> countSmaller(int[] nums) {
        List<Integer> result = new ArrayList<>();
		if (nums == null || nums.length == 0)
            return result;
		TreeNode root = new TreeNode(nums[nums.length - 1]);
		result.add(0);
		for (int i = nums.length - 2; i >= 0; i--) {
			int count = insertNode(root, nums[i]);
			result.add(count);
		}
		Collections.reverse(result);
		return result;
    }

    private int insertNode(TreeNode root, int val) {
		int thisCount = 0;
		while (true) {
			if (val <= root.val) {
				root.count++;
				if (root.left == null) {
					root.left = new TreeNode(val);
                    break;
				} else {
					root = root.left;
				}
			} else {
				thisCount += root.count;
				if (root.right == null) {
					root.right = new TreeNode(val);
                    break;
				} else {
					root = root.right;
				}
			}
		}
		return thisCount;
	}
}
